### PR TITLE
openshift-resource/route/prometheus-rule-tester resource resolution and early-exit performance

### PR DESCRIPTION
**Pull request by geoberle**
this PR refactors the `openshift-resources`, `openshift-route` and `prometheus-rules-tester` to work with qontract-server resource resolution (instead of sequential fetching). the `openshift-resources` early-exit code has been refactored so it can be reused by the other mentioned integrations as well.

at the same time, general performance issues for early-exit have been adressed by:
* not parsing resource YAML during early exit - this is not required because the content of a resource is sufficient to drive a decision for early-exit
* jinja2 template compilation caching

(!) review commit by commit

depends on https://github.com/app-sre/qontract-schemas/pull/227
relates to relates to https://issues.redhat.com/browse/APPSRE-6098

### CODE REVIEW

File: reconcile/openshift_resources_base.py [hunk 1]
```diff
@@ -234,11 +248,30 @@ def lookup_graphql_query_results(query: str, **kwargs) -> list[Any]:
...
+
+    jinja_env = jinja2.Environment(
+        extensions=[B64EncodeExtension, RaiseErrorExtension],
+        undefined=jinja2.StrictUndefined,
+        **env,
+    )
+    return jinja_env.from_string(body)
+
+
+def process_jinja2_template(body, vars=None, extra_curl: bool = False, settings=None):
```
Comment (rporres) [inline reconcile/openshift_resources_base.py:272]:
```suggestion
def process_jinja2_template(body, vars=None, extra_curly: bool = False, settings=None):
```

File: reconcile/openshift_resources_base.py [hunk 2]
```diff
@@ -256,12 +289,7 @@ def process_jinja2_template(body, vars=None, env=None, settings=None):
     vars.update({"query": lookup_graphql_query_results})
     vars.update({"url": url_makes_sense})
     try:
-        env = jinja2.Environment(
-            extensions=[B64EncodeExtension, RaiseErrorExtension],
-            undefined=jinja2.StrictUndefined,
-            **env,
-        )
-        template = env.from_string(body)
+        template = compile_jinja2_template(body, extra_curl)
```
Comment (rporres) [inline reconcile/openshift_resources_base.py:292]:
```suggestion
        template = compile_jinja2_template(body, extra_curly)
```

File: reconcile/openshift_resources_base.py [hunk 3]
```diff
@@ -234,11 +248,30 @@ def lookup_graphql_query_results(query: str, **kwargs) -> list[Any]:
     return results
 
 
-def process_jinja2_template(body, vars=None, env=None, settings=None):
+@lru_cache
```
Comment (rporres) [inline reconcile/openshift_resources_base.py:251]: I love this idea. Have you tried with `@cache`? Since we have already a lot of things in memory, we may want to add a few more with an unbound cache 😬

File: reconcile/openshift_resources_base.py [hunk 4]
```diff
@@ -817,3 +858,79 @@ def run(
         sys.exit(1)
 
     return ri
+
+
+def early_exit_desired_state(
```
Comment (rporres) [inline reconcile/openshift_resources_base.py:863]: This function is a complicated beast. I think it is a good moment to add a comment to guide the new reader of this.

File: reconcile/openshift_resources_base.py [hunk 5]
```diff
@@ -817,3 +858,79 @@ def run(
...
+def early_exit_monkey_patch_assign(
+    lookup_secret,
+    lookup_github_file_content,
+    url_makes_sense,
+    check_alertmanager_config,
+):
+    sys.modules[__name__].lookup_secret = lookup_secret
+    sys.modules[__name__].lookup_github_file_content = lookup_github_file_content
+    sys.modules[__name__].url_makes_sense = url_makes_sense
+    sys.modules[__name__].check_alertmanager_config = check_alertmanager_config
```
Comment (rporres) [inline reconcile/openshift_resources_base.py:936]: why this change?

File: reconcile/openshift_resources_base.py [hunk 6]
```diff
@@ -769,6 +809,7 @@ def get_namespaces(
     providers: Optional[list[str]] = None,
     cluster_name: Optional[str] = None,
     namespace_name: Optional[str] = None,
+    resource_schema_filter: Optional[str] = None,
```
Comment (rporres) [inline reconcile/openshift_resources_base.py:812]: I love this <3

File: reconcile/openshift_resources_base.py (hunk 5 above)
Comment (geoberle) [reply reconcile/openshift_resources_base.py:936]: the original code looked like
```
orb.lookup_secret = lookup_secret
```
since we are now in the `orb` module itself, it needed to be rewritten. so i tried
```
lookup_secret = patched_lookup_secret
```
but then vscode syntax highlighting was broken. it did not show, that `lookup_secret` would be the module function that is currently being overwritten. this might be just me but i rely heavily on syntax and color highlighting to get a feeling what is going on.

so i decided to be very explicit and that is how i landed with 

```
sys.modules[__name__].lookup_secret = lookup_secret
```

i'm open for suggestions

File: reconcile/openshift_resources_base.py (hunk 1 above)
Comment (geoberle) [reply reconcile/openshift_resources_base.py]: fixed

File: reconcile/openshift_resources_base.py (hunk 3 above)
Comment (geoberle) [reply reconcile/openshift_resources_base.py:251]: fixed

File: reconcile/openshift_resources_base.py (hunk 2 above)
Comment (geoberle) [reply reconcile/openshift_resources_base.py]: fixed

File: reconcile/openshift_resources_base.py (hunk 3 above)
Comment (geoberle) [reply reconcile/openshift_resources_base.py:251]: i did not consider an unbounded cache until now. thinking about this, i'm not too much concerned about this cache filling up uncontrolled. we don't have that many template changes that we would be hit by this. we restart reconciler way more often than we change templates.

additionally, cache eviction with lru_cache might cost performance gains. so i'm in favour of your proposal!

Comment (geoberle) [reply reconcile/openshift_resources_base.py]: changed in [7a13b0d](https://github.com/app-sre/qontract-reconcile/pull/2734/commits/7a13b0d87d07f4d90205529e58852b580805d24a)

File: reconcile/openshift_resources_base.py (hunk 5 above)
Comment (rporres) [reply reconcile/openshift_resources_base.py:934]: Fair enough :)

File: reconcile/openshift_resources_base.py (hunk 4 above)
Comment (geoberle) [reply reconcile/openshift_resources_base.py:863]: added comments in [c22bed7](https://github.com/app-sre/qontract-reconcile/pull/2734/commits/c22bed7b258faa8f4afd65c148e1d8377a64bc8a)

Comment (rporres) [reply reconcile/openshift_resources_base.py:863]: wunderbar! ❤️

**Review by rporres — APPROVED** [review]
